samples/core/callback/main.c

```diff
@@ -0,0 +1,517 @@
+    fill_color.y = 1.F;
+    fill_origin[0] = cb_opts.side / 2;
+    fill_origin[1] = cb_opts.side / 2;
+    fill_region[0] = cb_opts.side / 100;
+    fill_region[1] = cb_opts.side / 100;
+    OCLERROR_RET(clEnqueueFillImage(state.compute_queue, state.img_a,
+                                    &fill_color, fill_origin, fill_region, 0,
+                                    NULL, NULL),
+                 error, free_image_b);
+    OCLERROR_PAR(state.copy_buffer = clCreateBuffer(
+                     state.context, CL_MEM_READ_WRITE,
+                     cb_opts.side * cb_opts.side * sizeof(cl_uchar4), NULL,
+                     &error),
+                 error, free_image_b);
+
+    state.compute_event = NULL;
+    state.copy_event = NULL;
+    state.read_event = NULL;
+    for (size_t iter = 0; iter < cb_opts.iterations; ++iter)
+    {
+        cl_uint work_dim = 2;
+        size_t global_work_size[2] = { cb_opts.side, cb_opts.side };
+        OCLERROR_RET(
+            clSetKernelArg(state.kernel, 0, sizeof(cl_mem), &state.img_a),
+            error, copy_buf);
+        OCLERROR_RET(
+            clSetKernelArg(state.kernel, 1, sizeof(cl_mem), &state.img_b),
+            error, copy_buf);
+        if (state.compute_event)
+        {
+            OCLERROR_RET(clReleaseEvent(state.compute_event), error, copy_buf);
+        }
+        OCLERROR_RET(clEnqueueNDRangeKernel(
+                         state.compute_queue, state.kernel, work_dim, NULL,
+                         global_work_size, NULL,
+                         state.read_event == NULL ? 0 : 1,
+                         state.read_event == NULL ? NULL : &state.read_event,
+                         &state.compute_event),
+                     error, copy_buf);
+
+        if (iter % cb_opts.write_iter == 0)
+        {
+            size_t copy_origin[3] = { 0, 0, 0 };
+            size_t copy_region[3] = { cb_opts.side, cb_opts.side, 1 };
+            if (state.copy_event)
+            {
+                OCLERROR_RET(clReleaseEvent(state.copy_event), error,
+                             compute_ev);
+            }
+            OCLERROR_RET(clEnqueueCopyImageToBuffer(
+                             state.copy_queue, state.img_a, state.copy_buffer,
+                             copy_origin, copy_region, 0, 1u,
+                             &state.compute_event, &state.copy_event),
+                         error, compute_ev);
+            if (state.read_event)
+            {
+                OCLERROR_RET(clReleaseEvent(state.read_event), error, copy_ev);
+            }
+            OCLERROR_RET(
+                enqueue_buffer_read(state.read_queue, state.copy_buffer,
+                                    cb_opts.side, iter / cb_opts.write_iter,
+                                    &state.copy_event, &state.read_event),
+                error, copy_ev);
+        }
+
+        cl_mem tmp = state.img_a;
+        state.img_a = state.img_b;
+        state.img_b = tmp;
+    }
+
```

Comment:
I'm not exactly sure that it's at this point in the program, but the C++ version has a call to `clFinish` around here.  Do we need this in the C version also?

```c++
        // Wait for every read finished callback to fire.
        read_queue.finish();
```

Reply:
Very observant finding, thank you!

Shortly: no, it is not necessary, since in the C example, the wait loop in the end of `main` is slightly different from the C++ version.
- In the C++ version, before calling `.wait()` on every write image task, we must be sure that every element of the `image_write_tasks` contains a valid task item. This is ensured by calling `read_queue.finish();`, that synchronizes with the execution of the `read_complete_callback`s that create the entries in `image_write_tasks`.
- Whereas in the C version, we wait for `g_state.image_save_finished_count` to reach a predetermined value. Thereby it won't hurt if we enter the critical section *before* all the images were saved.

Reply:
Oof, what a pain.  This is why we can't have nice things :smile: 

Thanks for the explanation, I'll just resolve this conversation.

Reply:
I actually only noticed this because I ran these samples with debugging and profiling tools enabled, specifically the [OpenCL Intercept Layer](https://github.com/intel/opencl-intercept-layer/blob/main/docs/chrome_tracing.md) and I didn't see any activity on the device in the C version.  This is because the C sample currently does not make any blocking calls that would otherwise cause the device activity to be "flushed".  I don't think this is necessarily a problem in the sample, but it's definitely odd and not something I've seen before!  I'll probably need to add provisions to handle this in the OpenCL Intercept Layer.

My one thought though is that without any blocking calls, calls to `clFinish`, or even calls to `clFlush`, if an implementation is maximally "lazy" and buffers everything, is there anything that would cause execution to begin?  I haven't seen this in any of the implementations that I've tested but I think it's possible in theory.

Reply:
Thank you for mentioning this tool! I didn't know it before, but it looks to be very useful. I could reproduce the same thing: device timing was only available when a call to `clFinish` was made. `clFlush` was not enough.

For know, I've updated the C example with a `clFlush` to all 3 queues. Based on what you suggest and the API spec, this seems to be the correct thing to do. Let me know if you find this a satisfying resolution.

Reply:
Thanks! I think the calls to `clFlush` should be sufficient.

I also just pushed a fix to the tool for this so I think we should be good.

https://github.com/intel/opencl-intercept-layer/pull/326